### DISCUSSION TITLE
Is it (technically) possible to change ZigBee channel without having to re-pair every device?

### DISCUSSION THREAD

**Discussion by Aephir** [zigpy/zigpy#908] [Q&A]
I just switched from deCONZ to [ZHA](https://www.home-assistant.io/integrations/zha), and found that my ZigBee network was no longer on channel 25 (as I had set up in deCONZ), but instead defaulted to channel 15. 

Since the ZigBee network with ZHA was very unreliable, I wanted to switch back to channel 25 (which is as good as free from WiFi interference). But after having re-paired ~60 devices for moving to ZHA, I was daunted by the information I found, that changing channel would mean that I'd have to go through that again. Re-pairing every device.

Is that a technical limitation? Or is it something that could be circumvented, so a "quick channel switch" could be possible in the future?

**Comment by Hedda** [comment]
FYI, the question was brought up before in https://github.com/zigpy/zigpy/issues/573

I have not tried myself but if I interpret Hubutat docs and some of these posts in their community at least some devices (but not all) are capable of automatically moving to a new channel eventually, (and by eventually I am assuming that it can probably take up to 24-hours), however, some devices will require manual intervention, such as pressing a button or reseating battery to wake them up or more, or likely, re-pairing.

https://docs.hubitat.com/index.php?title=Zigbee_Details

"_Zigbee Channel (B) - Although your hub will automatically select a Zigbee channel for operation during initial setup. you may find in some cases that it will be necessary to manually change the Zigbee channel, as not all devices will operate properly on the higher channels. Consult you device manufacturer instructions for details on the channel ideally suited for your device. You may find that even when all of the devices currently paired to you hub are operating properly, a single device you wish to add may require that you change the channel for it to pair or operate properly. If you do need to change your Zigbee channel, please be patient, as it **may take up to 24 hours or more** for some devices to update to the new channel._"

https://community.home-assistant.io/t/changing-the-zigbee-channel/322789

https://community.hubitat.com/t/question-have-you-changed-the-zigbee-channel/76388/

So sounds as if changing the Zigbee channel without re-pairing might not actually be a technical limitation of the Zigbee specification, or?

Maybe it depends on which Zigbee stack? Or version of Zigbee Coordinator firmware/SDK? Or Zigbee Coordinator firmware configuration?

There is some more info on it for ZHA from Adminiuga and martin3000 in https://github.com/home-assistant/home-assistant.io/pull/18865

Summery; Zigbee channel is set when initially forming the Zigbee network (when the network is initialized), i.e. a ZHA/zigpy limitation?

Maybe the zigpy wiki and the Home Assistant's ZHA integration end-user documentation needs to be made more clear why this is so? 

https://github.com/zigpy/zigpy/wiki/Zigbee---Changing-channel

https://www.home-assistant.io/integrations/zha#defining-zigbee-channel-to-use

"_ZHA prefers to use Zigbee channel 15 by default. You can change this using YAML configuration, but this only works if there’s no existing network. To change the channel for an existing network, radio has to be factory reset and a new network to be formed. This requires re-pairing of all the devices._"

Zigbee2MQTT (Z2M) also notes that changing the Zigbee channel requires re-pairing of all devices:

https://www.zigbee2mqtt.io/guide/configuration/zigbee-network.html#permit-join

```
# Optional: ZigBee channel, changing requires re-pairing of all devices. (Note: use a ZLL channel: 11, 15, 20, or 25 to avoid Problems)
```
```
# Note: changing requires repairing of all devices (default: shown below)
```

Homey Zigbee developers have posted a similar explanation here:

https://community.homey.app/t/changing-zigbee-channel/39985/3

**Reply by Hedda** [reply]
> So sounds as if changing the Zigbee channel without re-pairing might not actually be a technical limitation of the Zigbee specification, or?
> 
> Maybe it depends on which Zigbee stack? Or version of Zigbee Coordinator firmware/SDK? Or Zigbee Coordinator firmware configuration?

@Aephir If you are willing to try then please go ahead and test it and post result with details on your setup and which if any devices worked?

**Reply by MattWestb** [reply]
I have doing it one time with `zha_custom` and one Billy EZSP and it was working with all IKEA routers but as expected not with Aqara / Xiaomi sensors that need little help.
If all device is Zifbee 3 compatible it shall doing all as written in `docs-13-0402-13-00zi-Base-Device-Behavior-Specification-2-1.pdf` it shall not being any problem then one Zigbee mesh shall have the possibility scanning for interference and moving to one better channel but as long we doing with Xiaomi devices is no go.
Also rolling the network key or regular basis shall being dome and all devices shall flowing with if they is working as standard.

**Comment by Hedda** [comment]
@MattWestb linked to https://github.com/zigpy/zigpy-znp/issues/31 where he mention in https://github.com/zigpy/zigpy-znp/issues/31#issuecomment-691651792 that be tested changing the channel with when using ZHA with bellows (Silicon Labs EZSP adapter) and that it worked with an IKEA Trådfri device.

 @walthowd @TheJulianJES maybe also tried as edited zigpy wiki article https://github.com/zigpy/zigpy/wiki/Zigbee---Changing-channel ?

**Reply by TheJulianJES** [reply]
Late answer but puddly made a [script](https://dpaste.org/KOAHo) at one point that sends a command to let devices know of the channel change (and also changes the network channel of course). Here's the link to the Discord thread: https://discord.com/channels/330944238910963714/551843655682490399/threads/985585519880704041

I used that script successfully months ago but I think I needed to modify it slightly.
It worked for mains-powered devices and *some* end-devices. Most needed to be re-paired though.

**Comment by Hedda** [comment]
So only changing Zigbee channel for ZHA inside Home Assistant does not work for ConBee/RaspBee today as ZHA does not appear to save the new configuration to the Zigbee Coordinator, but with a workaround, it should be possible to change the channel while still using ZHA. 

Because changing Zigbee channel does not work in ZHA for ConBee/RaspBee today wizbang wrote a guide with a workaround which does change Zigbee channel for ZHA if got ConBee/RaspBee by using Phoscon deCONZ software on another computer just to change channels:

https://community.home-assistant.io/t/changing-the-channel-of-a-conbee-ii-zigbee-controller/328763

"**_For those of you who are also struggling to change channel. Here is what I did and I hope it helps._**

**_I run HA on a Synology NAS in a docker container. Using this to do anything clever with the ConBee II device is not advised as the marthoc/deconz docker container seems flakey at best when accessing the USB device,_**

**_I didn’t want to install any 3rd party software directly on to the NAS, so I decided to plug the ConBee II into a Windows machine and configure it directly using the DeConz software._**

**_The advice from DeConz was as follows:_**

- **_Open the DeConz software and connect to the ConBee II_**
- **_Open the Network Settings_**
  - **_Set the channel you want_** 
  - **_Click Save_**
  - _**Click Done**_
-**_Leave the Zigbee network by clicking the Leave button_**
- **_Exit the application_**
- **_To check the setting has been changed:_**
- **_Unplug and plug back in the COnBee II_**
- **_Restart DeConz and enter the Network Settings dialog_**
  - **_Click Read_**
  - **_Check channel is selected_**
- **_Exit the software_**
- **_Unplug your ConBee II from the Windows machine and plug it back into your NAS_**
- **_Restart Home Assistant_**

**_HA will now use the new channel, but you will need to wait for all your Zigbee devices to time out on the old channel and scan for the new channel. I found some devices switched to the new channel in 10-15mins, but some took over 4 hours to come back online in HA. Be patient. They all came back online for me in the end_**"

Read some users reporting that only had to manually wake Zigbee end-point devices (i.e. battery operated Zigbee devices that only change state change needs to be forced to send new state to initiate communication and in which chase they will detect new channel and change it if they support it. So for example open door/window if got a door/window sensor).

So in conclusion it sounds as if this problem of not being able to change channel should be reported as an issue for ZHA or zigpy-deconz?

https://github.com/home-assistant/core/issues

https://github.com/zigpy/zigpy-deconz/issues

For reference; apparently, the later Phoscon deCONZ software for ConBee/RaspBee look to even got a GUI option for changing Zigbee channel so obviously, it gives end-users an easy option to change channels without re-pairing that works when running deCONZ implementation. 

![image](https://user-images.githubusercontent.com/6320001/153826853-f641bb28-e2a5-44c0-b95e-a0ba780d7e9c.png)

PS: @Aephir Best if you can change the "**_Category_**" of this topic from "**_Q&A_**" to "**_ZHA development_**" since should really only apply to ZHA.

**Reply by Hedda** [reply]
Note! By the way, sounds like might be a bug in some RaspBee/ConBee firmware version that prevent changing to above channel 15:

https://github.com/dresden-elektronik/deconz-rest-plugin/issues/5169

https://forum.phoscon.de/t/how-do-i-change-a-conbee-ii-channel/302

**Comment by Hedda** [comment]
> Is that a technical limitation? Or is it something that could be circumvented, so a "quick channel switch" could be possible in the future?

FYI, I think @puddly now answered the original question specifically in regards to zigpy and ZHA implementations here -> https://github.com/zigpy/zigpy/issues/842#issuecomment-1039443592

> > _Could maybe `set_channel` be made a standard method to allow changing Zigbee channels too?_
> 
> _**You can already change the network channel with `write_network_info`. Changing the coordinator's channel alone won't do very much, though.**_
> 
> _**If this is going to be done, it'll have to be with `zdo.Mgmt_NWK_Update_req` and will need to be implemented within zigpy, not the radio libraries. It's not a simple task. The only way I can see it working is by using the network topology to traverse the existing devices depth-first and migrate them one-by-one with unicast requests.**_

**Comment by pipiche38** [comment]
On the ZigbeeforDomoticz plugin we are doing it (with the zigpy layer and in native zigate as well). If your end device is ZB3.0 compliant, you will change the channel, notified all listening devices and they will shift to the new channel, and then the ZB3.0 ZED will send a device annoucement and will takle the new channel

**Reply by MattWestb** [reply]
Changing channel and also roiling network key is one part of the `Base-Device-Behavior-Specification` but many SED is not following the spec like all old Xaomi / aqara sensors is no real ZB3 devices (and some other manufacture is worse only writing Zigbee on the package and no certificated device) and is not expected doing the "move" OK.
